python/tank/bootstrap/async.py

```diff
@@ -0,0 +1,278 @@
+        """
+
+        # Make the QThread object exit its event loop.
+        self._thread.quit()
+
+        # Make the worker operate with the main thread affinity
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        try:
+
+            # Ladies and Gentlemen, start your engines!
+            engine = self._toolkit_manager._start_engine(toolkit, self._engine_name, self._entity)
+
+        except Exception, exception:
+
+            if self._failed_callback:
+                # Handle cleanup after failed completion of the engine startup.
+                self._failed_callback(self._toolkit_manager.ENGINE_STARTUP_PHASE, exception)
+
+            return
+
+        if self._completed_callback:
+            # Handle cleanup after successful completion of the engine startup.
+            self._completed_callback(engine)
+
+    @QtCore.Slot(Exception)
+    def _fail_bootstrap(self, exception):
+        """
+        Callback slot that handles cleanup after failed completion of the toolkit bootstrap.
+
+        :param exception: Exception raised while bootstrapping the toolkit.
+        """
+
+        # Make the QThread object exit its event loop.
+        self._thread.quit()
+
+        # Make the worker operate with the main thread affinity
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        if self._failed_callback:
+            # Handle cleanup after failed completion of the toolkit bootstrap.
+            self._failed_callback(self._toolkit_manager.TOOLKIT_BOOTSTRAP_PHASE, exception)
+
+
+class _BootstrapToolkitWorker(QtCore.QObject):
+    """
+    Bootstrap worker that can bootstrap an :class:`~sgtk.Sgtk` instance asynchronously in a background thread.
+
+    This class defines 3 Qt signals:
+    - ``progressing`` with signature ``(float, str, int, int)``
+      emitted while the bootstrap toolkit worker is progressing in its work in the background,
+      where ``float, str, int, int``are the current progress value, the progress message to report,
+      the current item number being looped over and the maximum item number being looped over.
+    - ``completed`` with signature ``(Sgtk)``
+      emitted when the bootstrap toolkit worker successfully completes its work in the background,
+      where ``Sgtk`` is the bootstrapped :class:`~sgtk.Sgtk` instance.
+    - ``failed`` with signature ``(Exception)``
+      emitted when the bootstrap toolkit worker fails to complete its work in the background,
+      where ``Exception`` is the python exception raised while bootstrapping.
+    """
+
+    # Qt signal emitted while the bootstrap toolkit worker is progressing in its work in the background.
+    progressing = QtCore.Signal(float, str, int, int)
+
+    # Qt signal emitted when the bootstrap toolkit worker successfully completes its work in the background.
+    completed = QtCore.Signal(Sgtk)
```

Comment:
in my experience, qt signals can be brittle. It may well be unnecessary, but I think in order to maximise stability (especially with DCCs like nuke running old versions of pyside), we should change this signal to not pass sgtk, 
and instead add an accessor - something like `get_sgtk()`.

```
completed = QtCore.Signal()

...
# in calling code
self._worker.completed.connect(self._complete_bootstrap)

def _complete_bootstrap(self):
   sgtk = self._worker.get_sgtk()
   ....
```